### PR TITLE
plugins/ollama: init

### LINKED ISSUE #778: Add support for ollama.nvim
**Plugin**: ollama.nvim
**Plugin repository**: https://github.com/nomnivore/ollama.nvim
**Is the plugin on nixpkgs**: No

## Extra information

Ollama server is available in nixpkgs https://github.com/NixOS/nixpkgs/blob/nixos-23.11/pkgs/tools/misc/ollama/default.nix#L43

### PR DESCRIPTION
**Pull request by GaetanLepage**
Add support for [ollama.nvim](https://github.com/nomnivore/ollama.nvim).

Fixes #778

### CODE REVIEW

File: plugins/utils/ollama.nix [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            ''
+              A `string.match` pattern to use for an Action to extract the output from the response
+              (Insert/Replace).
+            '';
+
+          options = helpers.mkNullOrOption (with types; attrsOf anything) ''
+            Additional model parameters, such as temperature, listed in the documentation for the [Modelfile](https://github.com/jmorganca/ollama/blob/main/docs/modelfile.md#valid-parameters-and-values).
+          '';
+
+          system = helpers.mkNullOrOption types.str ''
```
Comment (traxys) [inline plugins/utils/ollama.nix:125]: Please allow a rawType with `helpers.maybeRaw`

**Review by traxys — CHANGES_REQUESTED** [review]

File: plugins/utils/ollama.nix (hunk 1 above)
Comment (GaetanLepage) [reply plugins/utils/ollama.nix:125]: `helpers.maybeRaw` is not a thing. It is not exposed by the helpers but only used internally in `defaultNullOpts.*`.

Comment (traxys) [reply plugins/utils/ollama.nix:125]: Then we shoule either expose the maybeRaw, or put an either here

Comment (GaetanLepage) [reply plugins/utils/ollama.nix]: Done (in a separate commit)

File: lib/helpers.nix [hunk 2]
```diff
@@ -140,9 +140,9 @@ with lib; rec {
       ''
     );
 
-  defaultNullOpts = let
-    maybeRaw = t: lib.types.either t rawType;
-  in rec {
+  maybeRaw = t: lib.types.either t rawType;
+
+  defaultNullOpts = rec {
```
Comment (traxys) [inline lib/helpers.nix:145]: maybe put that in the newly created `nixvimTypes`

**Review by traxys — CHANGES_REQUESTED** [review]

File: lib/helpers.nix (hunk 2 above)
Comment (GaetanLepage) [reply lib/helpers.nix]: Like so ?

File: lib/helpers.nix [hunk 3]
```diff
@@ -100,9 +100,12 @@ with lib; rec {
       ''
     );
 
-  defaultNullOpts = let
-    maybeRaw = t: lib.types.either t nixvimTypes.rawLua;
-  in rec {
+  maybeRaw = type:
+    types.either
+    type
+    nixvimTypes.rawLua;
```
Comment (traxys) [inline lib/helpers.nix:106]: no, put `maybeRaw` in the `nixvimTypes` attrset

**Review by traxys — CHANGES_REQUESTED** [review]

File: plugins/utils/ollama.nix [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+      package = helpers.mkPackageOption "ollama.nvim" pkgs.vimPlugins.ollama-nvim;
+
+      model = helpers.defaultNullOpts.mkStr "mistral" ''
+        The default model to use.
+      '';
+
+      prompts = let
+        promptOptions = {
+          prompt = mkOption {
+            type = types.str;
```
Comment (traxys) [inline plugins/utils/ollama.nix:74]: maybeRaw types.str, if you want to dynamically create the prompt

File: plugins/utils/ollama.nix [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+      rawLua
+      (enum [
+        "display"
+        "replace"
+        "insert"
+        "display_replace"
+        "display_insert"
+        "display_prompt"
+      ])
+      (submodule {
```
Comment (traxys) [inline plugins/utils/ollama.nix:22]: The submodule options won't be rendered in the documentation I think? I'm not sure if there is a way to fix that :/

File: plugins/utils/ollama.nix [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+          model = helpers.mkNullOrOption types.str ''
+            The model to use for this prompt.
+
+            Defaults to the value of `plugins.ollama.model`.
+          '';
+
+          extract =
+            helpers.defaultNullOpts.mkNullable
+            (with types; either str (enum [false]))
```
Comment (traxys) [inline plugins/utils/ollama.nix:109]: I think this needs a `maybeRaw`

File: plugins/utils/ollama.nix [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+          action = helpers.mkNullOrOption actionOptionType ''
+            How to handle the output.
+
+            See [here](https://github.com/nomnivore/ollama.nvim/tree/main#actions) for more details.
+
+            Defaults to the value of `plugins.ollama.action`.
+          '';
+
+          model = helpers.mkNullOrOption types.str ''
```
Comment (traxys) [inline plugins/utils/ollama.nix:101]: This needs a `maybeRaw`

File: plugins/utils/ollama.nix [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+              format
+              ;
+          }
+          else prompt;
+      in
+        mkOption {
+          type = with types;
+            attrsOf (
+              either
+              (submodule {
```
Comment (traxys) [inline plugins/utils/ollama.nix:153]: I think doing the either means the documentation won't render the prompt options :/

**Review by traxys — CHANGES_REQUESTED** [review]

File: plugins/utils/ollama.nix (hunk 6 above)
Comment (GaetanLepage) [reply plugins/utils/ollama.nix:109]: `helpers.defaultNullOpts.mkStrLuaFnOr` is even better !

File: plugins/utils/ollama.nix (hunk 8 above)
Comment (GaetanLepage) [reply plugins/utils/ollama.nix:153]: Maybe, but we don't really have the choice here...

File: plugins/utils/ollama.nix (hunk 4 above)
Comment (GaetanLepage) [reply plugins/utils/ollama.nix]: Indeed !

File: plugins/utils/ollama.nix (hunk 7 above)
Comment (GaetanLepage) [reply plugins/utils/ollama.nix]: Done

File: plugins/utils/ollama.nix (hunk 5 above)
Comment (GaetanLepage) [reply plugins/utils/ollama.nix:22]: No idea...

**Review by traxys — APPROVED** [review]

**Pull request merged**